javascript/ql/lib/semmle/javascript/security/dataflow/TaintedPathCustomizations.qll

```diff
@@ -109,470 +122,552 @@ module TaintedPath {
         not (this.isNormalized() and this.isAbsolute())
       }
     }
-
-    /**
-     * A flow label representing an array of path elements that may include "..".
-     */
-    abstract class SplitPath extends DataFlow::FlowLabel {
-      SplitPath() { this = "splitPath" }
-    }
   }
 
   /**
-   * Holds if `s` is a relative path.
+   * Holds if `node` is a prefix of the string `../`.
    */
-  bindingset[s]
-  predicate isRelative(string s) { not s.charAt(0) = "/" }
+  private predicate isDotDotSlashPrefix(DataFlow::Node node) {
+    node.getStringValue() + any(string s) = "../"
+    or
+    // ".." + path.sep
+    exists(StringOps::Concatenation conc | node = conc |
+      conc.getOperand(0).getStringValue() = ".." and
+      conc.getOperand(1).getALocalSource() = NodeJSLib::Path::moduleMember("sep") and
+      conc.getNumOperand() = 2
+    )
+  }
 
   /**
-   * A call that normalizes a path.
+   * Provides logic for tracking taint flow through `DataFlow::CallNode`s
+   * that return paths.
    */
-  class NormalizingPathCall extends DataFlow::CallNode {
-    DataFlow::Node input;
-    DataFlow::Node output;
+  module PathNodes {
+    /** An abstract description of what kind of paths a call may return. */
+    abstract class PathProducingNode extends DataFlow::CallNode {
+      DataFlow::Node input;
+      DataFlow::Node output;
 
-    NormalizingPathCall() {
-      this = NodeJSLib::Path::moduleMember("normalize").getACall() and
-      input = this.getArgument(0) and
-      output = this
-    }
+      PathProducingNode() { this = this }
 
-    /**
-     * Gets the input path to be normalized.
-     */
-    DataFlow::Node getInput() { result = input }
+      abstract predicate mightReturnPath(
+        Label::PosixPath inputlabel, DataFlow::Node inputNode, Label::PosixPath outputlabel,
+        DataFlow::Node outputNode
+      );
 
-    /**
-     * Gets the normalized path.
-     */
-    DataFlow::Node getOutput() { result = output }
-  }
+      DataFlow::Node getInput() { result = input }
 
-  /**
-   * A call that converts a path to an absolute normalized path.
-   */
-  class ResolvingPathCall extends DataFlow::CallNode {
-    DataFlow::Node input;
-    DataFlow::Node output;
-
-    ResolvingPathCall() {
-      this = NodeJSLib::Path::moduleMember("resolve").getACall() and
-      input = this.getAnArgument() and
-      output = this
-      or
-      this = NodeJSLib::FS::moduleMember("realpathSync").getACall() and
-      input = this.getArgument(0) and
-      output = this
-      or
-      this = NodeJSLib::FS::moduleMember("realpath").getACall() and
-      input = this.getArgument(0) and
-      output = this.getCallback(1).getParameter(1)
+      DataFlow::Node getOutput() { result = output }
```

Comment:
I remember how `getInput` and `getOutput` was used previously.  
(In the `step` predicate).  

Now you have the `mightReturnPath` predicate, so I don't see what the `getInput()`/`getOutput()` predicates are used for.   
I can see that you use the predicates in some places where you use one transformer to define another, but in those cases you could define `getInput`/`getOutput` on those classes. 

Some of the implementations of `PathProducingNode` could also be simpler by not using the `input`/`output` fields.  
(E.g. in `RelativePathCall` you could trivially remove `input`/`output`).  

I think you should delete the fields from this class and the `getInput()` / `getOutput()` methods, and then only introduce them in the implementation-classes where they are needed.

Reply:
This was good advice, code is cleaner now and still does the same thing.

Reply:
👍